File: src/openrave.cpp
// -*- coding: utf-8 -*-
// Copyright (C) 2006-2009 Rosen Diankov (rdiankov@cs.cmu.edu)
//
// This file is part of OpenRAVE.
// OpenRAVE is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#endif

#include "libopenrave-core/openrave-core.h"
#include <set>

using namespace OpenRAVE;
using namespace std;

#include <stdio.h>
#include <boost/shared_ptr.hpp>
#include <boost/thread/thread.hpp>
#include <boost/array.hpp>
#include <boost/bind.hpp>
#include <boost/algorithm/string.hpp>

#include <signal.h>
void sigint_handler(int sig);
void MainOpenRAVEThread();

#ifdef _WIN32
#define usleep(micro) Sleep((micro)/1000)
#else
#define strnicmp strncasecmp
#define stricmp strcasecmp
#include <sys/time.h>
#endif

#define FORIT(it, v) for(it = (v).begin(); it != (v).end(); (it)++)

static bool bDisplayGUI = true, bShowGUI = true;

static EnvironmentBasePtr s_penv;
static boost::shared_ptr<boost::thread> s_mainThread;
static string s_sceneFile;
static string s_saveScene; // if not NULL, saves the scene and exits
static boost::shared_ptr<string> s_viewerName;

static list< pair<string, string> > s_listModules; // modules to initially create
static vector<string> vResourceFiles; // xml files to open
static int s_WindowWidth = 1024, s_WindowHeight = 768;
static int s_WindowPosX, s_WindowPosY;
static bool s_bSetWindowPosition = false;
int g_argc;
char** g_argv;
static bool s_bThreadDestroyed = false;
static const char* s_geometryextentsions[] = { "iv","vrml","wrl","stl","blend","3ds","ase","obj","ply","dxf","lwo","lxo","ac","ms3d","x","mesh.xml","irrmesh","irr","nff","off","raw"};
#ifndef _WIN32
#include <sys/stat.h>
#include <sys/types.h>
#endif

int main(int argc, char ** argv)
{
    g_argc = argc;
    g_argv = argv;

    int nServPort = 4765;

    DebugLevel debuglevel = Level_Info;
    list<string> listLoadPlugins;
    string collisionchecker, physicsengine, servername="textserver";
    bool bListPlugins = false;

    std::set<std::string> geometryextensions;
    for(size_t iext = 0; iext < sizeof(s_geometryextentsions)/sizeof(s_geometryextentsions[0]); ++iext) {
        geometryextensions.insert(s_geometryextentsions[iext]);
    }

    // parse command line arguments
    int i = 1;
    while(i < argc) {
        if((stricmp(argv[i], "-h") == 0)||(stricmp(argv[i], "-?") == 0)||(stricmp(argv[i], "/?") == 0)||(stricmp(argv[i], "--help") == 0)||(stricmp(argv[i], "-help") == 0)) {
            RAVELOG_INFO("OpenRAVE Usage\n"
                         "--nogui             Run without a GUI (does not initialize the graphics engine nor communicate with any window manager)\n"
                         "--hidegui           Run with a hidden GUI, this allows 3D rendering and images to be captured\n"
                         "--listplugins       List all plugins and the interfaces they provide\n"
                         "--loadplugin [path] load a plugin at the following path\n"
                         "--serverport [port] start up the server on a specific port (default is 4765)\n"
                         "--collision [name]  Default collision checker to use\n"
                         "--viewer [name]     Default viewer to use\n"
                         "--server [name]     Default server to use\n"
                         "--physics [name]    Default physics engine to use\n"
                         "-d [debug-level]   start up OpenRAVE with the specified debug level (higher numbers print more).\n"
                         "                   Default level is 2 for release builds, and 4 for debug builds.\n"
                         "-wdims [width] [height] start up the GUI window with these dimensions\n"
                         "-wpos x y set the position of the GUI window\n"
                         "--module [modulename] [args] Start openrave with a module. If args involves spaces, surround it with double quotes. args is optional.\n"
                         "--version          Output the current openrave version\n\n"
                         "-f [scene]         Load a openrave environment file\n\n");
            return 0;
        }
        else if((stricmp(argv[i], "--loadplugin") == 0)||(stricmp(argv[i], "-loadplugin") == 0)) {
            listLoadPlugins.push_back(argv[i+1]);
            i += 2;
        }
        else if((stricmp(argv[i], "--listplugins") == 0)||(stricmp(argv[i], "-listplugins") == 0)) {
            bListPlugins = true;
            i++;
        }
        else if( stricmp(argv[i], "--version") == 0 ) {
            printf("%s\n",OPENRAVE_VERSION_STRING);
            return 0;
        }
        else if( stricmp(argv[i], "-f") == 0 ) {
            s_sceneFile = argv[i+1];
            i += 2;
        }
        else if( stricmp(argv[i], "-save") == 0 ) {
            s_saveScene = argv[i+1];
            i += 2;
        }
        else if((stricmp(argv[i], "--collision") == 0)||(stricmp(argv[i], "-collision") == 0)) {
            collisionchecker = argv[i+1];
            i += 2;
        }
        else if((stricmp(argv[i], "--viewer") == 0)||(stricmp(argv[i], "-viewer") == 0)) {
            s_viewerName.reset(new string(argv[i+1]));
            i += 2;
        }
        else if((stricmp(argv[i], "--physics") == 0)||(stricmp(argv[i], "-physics") == 0)) {
            physicsengine = argv[i+1];
            i += 2;
        }
        else if((stricmp(argv[i],"--server") == 0)||(stricmp(argv[i],"-server") == 0)) {
            servername = argv[i+1];
            i += 2;
        }
        else if((stricmp(argv[i], "--module") == 0)||(stricmp(argv[i], "-problem") == 0)) {
            s_listModules.push_back(pair<string, string>(argv[i+1], ""));
            i += 2;

            if((i < argc)&&(argv[i][0] != '-')) {
                // set the args
                s_listModules.back().second = argv[i];
                i++;
            }
        }
        else if((stricmp(argv[i], "--nogui") == 0)||(stricmp(argv[i], "-nogui") == 0)) {
            bDisplayGUI = false;
            i++;
        }
        else if((stricmp(argv[i], "--hidegui") == 0)||(stricmp(argv[i], "-hidegui") == 0)) {
            bShowGUI = false;
            i++;
        }
        else if( stricmp(argv[i], "-d") == 0 ) {
            debuglevel = (DebugLevel)atoi(argv[i+1]);
            i += 2;
        }
        else if( stricmp(argv[i], "-wdims") == 0 ) {
            s_WindowWidth = atoi(argv[i+1]);
            s_WindowHeight = atoi(argv[i+2]);
            i += 3;
        }
        else if( stricmp(argv[i], "-wpos") == 0 ) {
            s_WindowPosX = atoi(argv[i+1]);
            s_WindowPosY = atoi(argv[i+2]);
            s_bSetWindowPosition = true;
            i += 3;
        }
        else if((stricmp(argv[i], "--serverport") == 0)||(stricmp(argv[i], "-serverport") == 0)) {
            nServPort = atoi(argv[i+1]);
            i += 2;
        }
        else {
            // try to load as xml files
            vResourceFiles.push_back(argv[i]);
            i++;
        }
    }

    RaveInitialize(true);

    // create environment and start a command-line controlled simulation
    RaveSetDebugLevel(debuglevel);
    s_penv = RaveCreateEnvironment();
    RaveSetDebugLevel(debuglevel);
    for(list<string>::iterator it = listLoadPlugins.begin(); it != listLoadPlugins.end(); ++it) {
        RaveLoadPlugin(*it);
    }

    // add a signal handler
    signal(SIGINT,sigint_handler); // control C

    if( bListPlugins ) {

        std::list< std::pair<std::string, PLUGININFO> > plugins;
        std::list< std::pair<std::string, PLUGININFO> >::iterator itplugin;
        RaveGetPluginInfo(plugins);

        // output all the plugins and exit
        vector<string>::const_iterator itnames;

        vector<string>::iterator itname;
        stringstream ss;

        ss << endl << "Number of plugins: " << plugins.size() << endl;

        stringstream buf;
        std::map<InterfaceType,std::string>::const_iterator ittype;
        FORIT(ittype,RaveGetInterfaceNamesMap()) {
            vector<string> names;
            FORIT(itplugin, plugins) {
                FORIT(itnames, itplugin->second.interfacenames[ittype->first]) {
                    buf.str("");
#ifdef _WIN32
                    buf << "  " << *itnames << " - " << itplugin->first;
#else
                    buf << "  " << ChangeTextColor(0,OPENRAVECOLOR_DEBUGLEVEL) << *itnames << ResetTextColor() << " - " << itplugin->first;
#endif
                    names.push_back(buf.str());
                }
            }
            sort(names.begin(), names.end());
#ifdef _WIN32
            ss << ittype->second << ": " << names.size() << endl;
#else
            ss << ChangeTextColor(0,OPENRAVECOLOR_WARNLEVEL) << ittype->second << ": " << ResetTextColor() << names.size() << endl;
#endif
            FORIT(itname, names) {
                ss << itname->c_str() << endl;
            }
        }
        printf("%s",ss.str().c_str());
        s_penv->Destroy();
        return 0;
    }

    if((servername.size() > 0)&&(nServPort > 0)) {
        ModuleBasePtr pserver = RaveCreateModule(s_penv, servername);
        if( !!pserver ) {
            stringstream ss;
            ss << nServPort;
            if( s_penv->AddModule(pserver,ss.str()) != 0 )
                RAVELOG_WARN("failed to load server\n");
        }
    }

    if( collisionchecker.size() > 0 ) {
        CollisionCheckerBasePtr p = RaveCreateCollisionChecker(s_penv, collisionchecker);
        if( !!p ) {
            s_penv->SetCollisionChecker(p);
        }
    }
    if( physicsengine.size() > 0 ) {
        PhysicsEngineBasePtr p = RaveCreatePhysicsEngine(s_penv, physicsengine);
        if( !!p ) {
            s_penv->SetPhysicsEngine(p);
        }
    }

    s_bThreadDestroyed = false;
    s_mainThread.reset(new boost::thread(boost::bind(MainOpenRAVEThread)));
    s_mainThread->join();
    s_penv.reset();
    RaveDestroy();
    return 0;
}

// use to control openrave
void MainOpenRAVEThread()
{
    EnvironmentBasePtr penv = s_penv; // need to do this since s_penv can be reset at any time
    ViewerBasePtr pviewer;
    if( bDisplayGUI && (!s_viewerName ||(s_viewerName->size()>0)) ) {
        // find a viewer
        if( !!s_viewerName &&(s_viewerName->size() > 0)) {
            pviewer = RaveCreateViewer(penv, *s_viewerName);
        }
        if( !pviewer ) {
            boost::array<string,1> viewer_prefs = { { "qtcoin"}};
            for(size_t i = 0; i < viewer_prefs.size(); ++i) {
                pviewer = RaveCreateViewer(penv, viewer_prefs[i]);
                if( !!pviewer ) {
                    break;
                }
            }
        }

        if( !pviewer ) { // take any viewer
            std::map<InterfaceType, std::vector<std::string> > interfacenames;
            RaveGetLoadedInterfaces(interfacenames);
            std::vector<std::string>::const_iterator itname;
            FORIT(itname, interfacenames[PT_Viewer]) {
                pviewer = RaveCreateViewer(penv, *itname);
                if( !!pviewer ) {
                    break;
                }
            }
        }

        if( !pviewer ) {
            RAVELOG_WARN("failed to find an OpenRAVE viewer.\n");
        }
        else {
            RAVELOG_DEBUG("using %s viewer\n", pviewer->GetXMLId().c_str());
            pviewer->SetSize(s_WindowWidth, s_WindowHeight);
            if( s_bSetWindowPosition ) {
                pviewer->Move(s_WindowPosX,s_WindowPosY);
            }
            penv->AddViewer(pviewer);
        }
    }

    {
        EnvironmentMutex::scoped_lock lock(penv->GetMutex());

        if( s_sceneFile.size() > 0 ) {
            penv->Load(s_sceneFile);
        }
        vector<string>::iterator it;
        FORIT(it, vResourceFiles) {
            penv->Load(*it);
        }
        list< pair<string, string> >::iterator itprob;
        FORIT(itprob, s_listModules) {
            ModuleBasePtr prob = RaveCreateModule(penv, itprob->first);
            if( !!prob ) {
                penv->AddModule(prob, itprob->second);
            }
        }

        if( s_saveScene.size() > 0 ) {
            penv->Save(s_saveScene);
            s_bThreadDestroyed = true;
            return;
        }
    }

    // need to keep a local pointer around to guarantee destruction order
    if( !!pviewer ) {
        pviewer->main(bShowGUI);
        s_bThreadDestroyed = true;
    }
    else {
        while(!s_bThreadDestroyed) {
            usleep(10000);
        }
    }
    if( !!penv ) {
        penv->Remove(pviewer);
        penv.reset();
        s_penv.reset();
        int iter = 0;
        while(pviewer.use_count() > 1 && iter < 200) {
            RAVELOG_DEBUG("viewer use count > 1, waiting for others to release viewer so can guarantee destruction in correct thread\n");
            usleep(10000);
            ++iter;
        }
    }
}

void sigint_handler(int sig)
{
    RaveDestroy();
    s_penv.reset();
#ifndef _WIN32
    // have to let the default sigint properly shutdown the program
    signal(SIGINT, SIG_DFL);
    kill(getpid(), SIGINT);
#endif
}

File: python/openrave.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright (C) 2009-2010 Rosen Diankov (rosen.diankov@gmail.com)
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#     http://www.apache.org/licenses/LICENSE-2.0
# 
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
from __future__ import with_statement # for python 2.5
__author__ = 'Rosen Diankov'
__copyright__ = 'Copyright (C) 2009-2010 Rosen Diankov (rosen.diankov@gmail.com)'
__license__ = 'Apache License, Version 2.0'

import sys,time
try:
    from openravepy import *
except ImportError:
    print "openravepy is not set into PYTHONPATH env variable, attempting to add"
    # use openrave-config to get the correct install path
    from subprocess import Popen, PIPE
    try:
        openravepy_path = Popen(['openrave-config','--python-dir'],stdout=PIPE).communicate()
        sys.path.append(openravepy_path[0].strip())
    except OSError:
        import platform
        if sys.platform.startswith('win') or platform.system().lower() == 'windows':
            # in windows so add the default openravepy installation
            sys.path.append("C:\\Program Files\\openrave\\share\\openrave")
    from openravepy import *

from numpy import *
from optparse import OptionParser

def vararg_callback(option, opt_str, value, parser):
    assert value is None
    value = parser.rargs[:]
    del parser.rargs[:len(value)]
    setattr(parser.values, option.dest, value)

if __name__ == "__main__":
    parser = OptionParser(description='OpenRAVE %s'%openravepy.__version__,version=openravepy.__version__)
    parser.add_option('--listplugins', action="store_true",dest='listplugins',default=False,
                      help='List all plugins and the interfaces they provide.')
    parser.add_option('--loadplugin', action="append",type='string',dest='loadplugins',default=[],
                      help='List all plugins and the interfaces they provide.')
    parser.add_option('--collision', action="store",type='string',dest='collision',default=None,
                      help='Default collision checker to use')
    parser.add_option('--physics', action="store",type='string',dest='physics',default=None,
                      help='physics engine to use (default=%default)')
    parser.add_option('--viewer', action="store",type='string',dest='viewer',default='qtcoin',
                      help='viewer to use (default=%default)' )
    parser.add_option('--server', action="store",type='string',dest='server',default=None,
                      help='server to use (default=%default).')
    parser.add_option('--serverport', action="store",type='int',dest='serverport',default=4765,
                      help='port to load server on (default=%default).')
    parser.add_option('--debug','-d', action="store",type='string',dest='debug',default=None,
                      help='Debug level')
    parser.add_option('--database', action="callback",callback=vararg_callback, dest='database',default=None,
                      help='If specified, the next arguments will be used to call a database generator from the openravepy.databases module. The first argument is used to find the database module. For example:     %s --database grasping --robot=robots/pr2-beta-sim.robot.xml'%(sys.argv[0]))
    parser.add_option('--example', action="callback",callback=vararg_callback, dest='example',default=None,
                      help='If specified, the next arguments will be used to call an example from the openravepy.examples module. The first argument is used to find the example moduel. For example:     %s --example graspplanning --scene=data/lab1.env.xml'%(sys.argv[0]))
    parser.add_option('--ipython', '-i',action="store_true",dest='ipython',default=False,
                      help='if true will drop into the ipython interpreter rather than spin')
    parser.add_option('--listexamples',action='store_true',dest='listexamples',default=False,
                      help='Lists the available core examples')
    parser.add_option('--listdatabases',action='store_true',dest='listdatabases',default=False,
                      help='Lists the available core database generators')
    (options, args) = parser.parse_args()
    if options.listdatabases:
        for name in dir(databases):
            if not name.startswith('__'):
                print name
        sys.exit(0)
    if options.listexamples:
        for name in dir(examples):
            if not name.startswith('__'):
                print name
        sys.exit(0)
    if options.database is not None:
        args = options.database[0].split() + options.database[1:] # the first arg might also include the options
        try:
            database=getattr(databases,args[0])
        except (AttributeError,IndexError):
            print 'bad database generator, current list of executable generators are:'
            for name in dir(databases):
                if not name.startswith('__'):
                    print ' ' + name
            sys.exit(1)
        database.run(args=args)
        sys.exit(0)
    if options.example is not None:
        args = options.example[0].split() + options.example[1:] # the first arg might also include the options
        try:
            example = getattr(examples,args[0])
        except (AttributeError,IndexError):
            print 'bad example, current list of executable examples are:'
            for name in dir(examples):
                if not name.startswith('__'):
                    print ' ' + name
            sys.exit(1)
        example.run(args=args)
        sys.exit(0)
    if options.debug is not None:
        for debuglevel in [DebugLevel.Fatal,DebugLevel.Error,DebugLevel.Warn,DebugLevel.Info,DebugLevel.Debug,DebugLevel.Verbose]:
            if (not options.debug.isdigit() and options.debug.lower() == debuglevel.name.lower()) or (options.debug.isdigit() and int(options.debug) == int(debuglevel)):
                RaveSetDebugLevel(debuglevel)
                break
    env = Environment()
    try:
        if options.listplugins:
            plugins = env.GetPluginInfo()
            interfacenames = dict()
            for name,type in PluginType.names.iteritems():
                interfacenames[type] = []
            for pluginname,info in plugins:
                for type,names in info.interfacenames:
                    interfacenames[type] += [(n,pluginname) for n in names]
            print 'Number of plugins: %d'%len(plugins)
            for type,names in interfacenames.iteritems():
                print '%s: %d'%(str(type),len(names))
                names.sort()
                for interfacename,pluginname in names:
                    print '  %s - %s'%(interfacename,pluginname)
            sys.exit(0)
        if options.collision:
            cc = env.CreateCollisionChecker(options.collision)
            if cc is not None:
                env.SetCollisionChecker(cc)
        if options.physics:
            ph = env.CreatePhysicsEngine(options.physics)
            if ph is not None:
                env.SetPhysicsEngine(ph)
        if options.server:
            sr = env.CreateProblem(options.server)
            if sr is not None:
                env.LoadProblem(sr,'%d'%options.serverport)
        if options.viewer:
            env.SetViewer(options.viewer)
        # load files after viewer is loaded since they may contain information about where to place the camera
        for arg in args:
            if arg.endswith('.xml') or arg.endswith('.dae'):
                env.Load(arg)
        if options.ipython:
            with env:
                robots=env.GetRobots()
                robot=None if len(robots) == 0 else robots[0]
            from IPython.Shell import IPShellEmbed
            ipshell = IPShellEmbed(argv='',banner = 'OpenRAVE Dropping into IPython',exit_msg = 'Leaving Interpreter and closing program.')
            ipshell(local_ns=locals())
            sys.exit(0)
        while True:
            time.sleep(0.01)
    finally:
        env.Destroy()
